// bs-cli/lib/question.js
// evo init 问题选择
const question = {

  module: {
    type: 'list',
    message: `Select install module`,
    name: 'module',
    choices: ['module'],
  },
  name: {
    type: 'input',
    message: function(answer) {
      return answer.module === 'module' ? 'Project name' : 'Plugin name';
    },
    name: 'name',
    default: function(answer) {
      return answer.module === 'module' ? 'evo-web' : answer.module;
    },
  },
  version: {
    when: function(answer) {
      return answer.module === 'module';
    },
    type: 'input',
    message: 'Version',
    name: 'version',
    default: '0.0.1',
  },
  port: {
    when: function(answer) {
      return answer.module === 'module';
    },
    type: 'input',
    message: 'Server port',
    name: 'port',
    default: 7001,
  },
  author: {
    when: function(answer) {
      return answer.module === 'module';
    },
    type: 'input',
    message: 'Author',
    name: 'author',
  },
  description: {
    when: function(answer) {
      return answer.module === 'module';
    },
    type: 'input',
    message: 'Description',
    name: 'description',
  },
};

exports = module.exports = question;
